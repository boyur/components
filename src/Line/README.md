# Line

## Usage

```js
import { Line } from '@urbica/components';

<Line
  data={data}
  x={d => d.key}
  y={d => d.value}
  style={style}
  xScale={xScale}
  yScale={yScale}
  defined={d => !!d.value}
/>
```
