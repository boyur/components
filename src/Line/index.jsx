import React, { PropTypes } from 'react';
import { path } from 'd3-path';
import { line } from 'd3-shape';

const Line = (props) => {
  const { data, defined, curve, style, xScale, yScale, x, y } = props;
  const context = path();

  const lineGenerator = line()
    .x(d => xScale(x(d)))
    .y(d => yScale(y(d)))
    .context(context);

  if (curve) lineGenerator.curve(curve);
  if (defined) lineGenerator.defined(defined);

  lineGenerator(data);

  return <path d={context.toString()} fill="none" style={style} />;
};

Line.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  curve: PropTypes.func,
  defined: PropTypes.func,
  x: PropTypes.func.isRequired,
  y: PropTypes.func.isRequired,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired,
  style: PropTypes.shape({
    stroke: PropTypes.string,
    strokeWidth: PropTypes.number,
    strokeOpacity: PropTypes.number,
    strokeLinecap: PropTypes.string,
    strokeLinejoin: PropTypes.string
  }).isRequired
};

Line.defaultProps = {
  curve: PropTypes.null,
  defined: PropTypes.null
};

export default Line;
