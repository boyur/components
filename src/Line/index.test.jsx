/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import renderer from 'react-test-renderer';
import { scaleLinear } from 'd3-scale';

import Line from './';

test('Line', () => {
  const width = 100;
  const height = 100;

  const data = [{ x: 0, y: 0 }, { x: 1, y: 1 }];

  const x = d => d.x;
  const y = d => d.y;

  const xScale = scaleLinear().range([0, width]).domain([0, 1]);
  const yScale = scaleLinear().range([height, 0]).domain([0, 1]);

  const style = { stroke: '#000' };

  const component = renderer.create(
    <Line
      data={data}
      x={x}
      y={y}
      xScale={xScale}
      yScale={yScale}
      style={style}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
