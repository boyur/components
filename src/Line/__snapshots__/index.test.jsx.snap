// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line 1`] = `
<path
  d="M0,100L100,0"
  fill="none"
  style={
    Object {
      "stroke": "#000",
    }
  }
/>
`;
